### ISSUE TITLE
Improvements to file logging behaviour

### ISSUE THREAD

**Issue by felipesanches** [googlefonts/fontbakery#971]
@davelab6 suggested, at the very end of fb-check-ttf.py, the following expected behaviour for the file logging mechanism:

```
    # each fix line should set a fix flag, and
    # if that flag is True by this point, only then write the file
    # and then say any further output regards fixed files, and
    # re-run the script on each fixed file with logging level = error
    # so no info-level log items are shown
```

**Comment by vmesel** [comment]
Is it easy to implement @felipesanches ? I want to give a try, but don't know if its inside my knowledge

**Comment by felipesanches** [comment]
I don't think this is a trivial issue. The behaviour described in this feature request seems to require a large set of changes in the code.

I think that a good way to approach the problem would be to make each test become a proper function so that it can be invoked multiple times.

**Comment by m4rc1e** [comment]
Just wrote a lengthy post about this in #891.

Basically, we should always try and fix in source files instead of hot fix.

**Comment by felipesanches** [comment]
thanks, @m4rc1e I agree. And I replied there as well.

**Comment by davelab6** [comment]
> I think that a good way to approach the problem would be to make each test become a proper function so that it can be invoked multiple times.

This is now complete, right?

**Comment by felipesanches** [comment]
>> I think that a good way to approach the problem would be to make each test become a proper function so that it can be invoked multiple times.

> This is now complete, right?

Yes, it is!

**Comment by davelab6** [comment]
@felipesanches is there anything more to do here?

**Comment by davelab6** [comment]
@felipesanches ping

**Comment by felipesanches** [comment]
The only remaining task here is:

> @davelab6 suggested, at the very end of fb-check-ttf.py, the following expected behaviour for the file logging mechanism:
> 
> each fix line should set a fix flag, and
> if that flag is True by this point, only then write the file
> and then say any further output regards fixed files, and
> re-run the script on each fixed file with logging level = error
> so no info-level log items are shown

**Comment by davelab6** [comment]
Does that task still make sense with the new modularity?

**Comment by davelab6** [comment]
After discussing offline, I'm reassigning to @m4rc1e since he has dealt with hotfixing recently

**Comment by felipesanches** [comment]
The feature as described in your original request means that it would re-check the font binaries after applying hotfixes. There's not a lot of hotfix routines in FB since, in the past, we decided to focus on making it a good quality reporting tool for the Google Fonts collection. And we achieved that with the collection-wide web dashboard.

I'm strongly supportive of the idea of trying to push the maximum of fixes to the upstream font projects (fixing the sources) and reducing the amount of last-minute hotfixing of binaries. But I understand that some amount of hotfixing may still be needed. This feature would only make sense in a hotfixing scenario and I agree with @davelab6 that @m4rc1e is probably in a better position to let us know how to better address this. Either by dropping this feature or by confirming this is really needed and useful for his workflow.

**Comment by davelab6** [comment]
@m4rc1e do you think this should be closed?

**Comment by m4rc1e** [comment]
@davelab6 I think the logging is now rather beautiful.

![screen shot 2017-07-31 at 09 56 50](https://user-images.githubusercontent.com/7525512/28770325-9edb352a-75d6-11e7-8b93-313cf63c356a.png)


I'd still like us to add #1471 though. :-)

**Comment by davelab6** [comment]
Yep I agree :)

**Comment by m4rc1e** [comment]
I think we should close this. The log is great.
